src/table.rs

```diff
@@ -0,0 +1,1792 @@
+    /// also call it for entries with a different hash. `eq` should only return
+    /// true for the desired entry, at which point the search is stopped.
+    ///
+    /// When mutating an entry, you should ensure that it still retains the same
+    /// hash value as when it was inserted, otherwise lookups of that entry may
+    /// fail to find it.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # #[cfg(feature = "nightly")]
+    /// # fn test() {
+    /// use ahash::AHasher;
+    /// use hashbrown::HashTable;
+    /// use std::hash::{BuildHasher, BuildHasherDefault};
+    ///
+    /// let mut table = HashTable::new();
+    /// let hasher = BuildHasherDefault::<AHasher>::default();
+    /// let hasher = |val: &_| hasher.hash_one(val);
+    /// table.insert_unchecked(hasher(&1), (1, "a"), |val| hasher(&val.0));
+    /// if let Some(val) = table.find_mut(hasher(&1), |val| val.0 == 1) {
+    ///     val.1 = "b";
+    /// }
+    /// assert_eq!(table.find(hasher(&1), |val| val.0 == 1), Some(&(1, "b")));
+    /// assert_eq!(table.find(hasher(&2), |val| val.0 == 2), None);
+    /// # }
+    /// # fn main() {
+    /// #     #[cfg(feature = "nightly")]
+    /// #     test()
+    /// # }
+    /// ```
+    pub fn find_mut(&mut self, hash: u64, eq: impl FnMut(&T) -> bool) -> Option<&mut T> {
+        self.table
+            .find(hash, eq)
+            .map(|bucket| unsafe { bucket.as_mut() })
+    }
+
+    /// Returns an `OccupiedEntry` for an entry in the table with the given hash
+    /// and which satisfies the equality function passed.
+    ///
+    /// This can be used to remove the entry from the table. Call
+    /// [`HashTable::entry`] instead if you wish to insert an entry if the
+    /// lookup fails.
+    ///
+    /// This method will call `eq` for all entries with the given hash, but may
+    /// also call it for entries with a different hash. `eq` should only return
+    /// true for the desired entry, at which point the search is stopped.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # #[cfg(feature = "nightly")]
+    /// # fn test() {
+    /// use ahash::AHasher;
+    /// use hashbrown::HashTable;
+    /// use std::hash::{BuildHasher, BuildHasherDefault};
+    ///
+    /// let mut table = HashTable::new();
+    /// let hasher = BuildHasherDefault::<AHasher>::default();
+    /// let hasher = |val: &_| hasher.hash_one(val);
+    /// table.insert_unchecked(hasher(&1), (1, "a"), |val| hasher(&val.0));
+    /// if let Some(entry) = table.find_entry(hasher(&1), |val| val.0 == 1) {
+    ///     entry.remove();
+    /// }
+    /// assert_eq!(table.find(hasher(&1), |val| val.0 == 1), None);
+    /// # }
+    /// # fn main() {
+    /// #     #[cfg(feature = "nightly")]
+    /// #     test()
+    /// # }
+    /// ```
+    pub fn find_entry(
+        &mut self,
+        hash: u64,
+        eq: impl FnMut(&T) -> bool,
+    ) -> Option<OccupiedEntry<'_, T, A>> {
```

Comment:
Limitations in the borrow checker make it hard to use `find_entry` to implement my own non-growing `entry`, because if I return my wrapper around this `OccupiedEntry`, I can't have any _other_ branch that tries to store `&mut self` (to allow vacant insert).

Maybe instead:

```suggestion
    ) -> Result<OccupiedEntry<'_, T, A>, &mut Self> {
```

Reply:
This is surprising. I'm pretty sure the borrow checker is able to tell that a value is no longer used in a branch. Can you show the cause that causes this?

Reply:
Consider something like this:

```rust
use hashbrown::hash_table::{HashTable, OccupiedEntry};

struct Map(HashTable<(i32, String)>);

enum Entry<'a> {
    Occupied(OccupiedEntry<'a, (i32, String)>, i32),
    Vacant(&'a mut Map, i32)
}

impl Map {
    fn entry(&mut self, key: i32) -> Entry<'_> {
        if let Some(e) = self.0.find_entry(0, |&(x, _)| x == key) {
            Entry::Occupied(e, key)
        } else {
            Entry::Vacant(self, key)
        }
    }
}

fn main() {
    let mut map = Map(HashTable::new());
    let _ = map.entry(42);
}
```

```
error[E0499]: cannot borrow `*self` as mutable more than once at a time
  --> examples/mymap.rs:15:27
   |
11 |     fn entry(&mut self, key: i32) -> Entry<'_> {
   |              - let's call the lifetime of this reference `'1`
12 |         if let Some(e) = self.0.find_entry(0, |&(x, _)| x == key) {
   |                          ---------------------------------------- first mutable borrow occurs here
13 |             Entry::Occupied(e, key)
   |             ----------------------- returning this value requires that `self.0` is borrowed for `'1`
14 |         } else {
15 |             Entry::Vacant(self, key)
   |                           ^^^^ second mutable borrow occurs here
```

This kind of conditional return is a known limitation of NLL, which Polonius may eventually resolve:
https://smallcultfollowing.com/babysteps/blog/2018/06/15/mir-based-borrow-check-nll-status-update/#polonius

Reply:
The known workaround is to use a double-lookup, like `if self.contains(..) { let e = self.0.find_entry(..).unwrap(); .. } else { .. }`, so NLL can split the return lifetime.

Reply:
May be you can use `HashTable::entry` directly?

Reply:
I don't want the growth, as `entry` documents:

```rust
    /// This method may grow the table in preparation for an insertion. Call
    /// [`HashTable::find_entry`] if this is undesirable.
```

Reply:
I made the change, but I'm not too happy about it. In particular, `Self` makes for a very poor error type.

Reply:
Alternatively, `VacantEntry` could hold an `Option<InsertSlot>` instead, possibly even setting that from `find_entry` if it can do so without growing. But the rest of the `Entry` and `VacantEntry` methods would need to add parameters to deal with the possibility of growing later.

Or we could add separate `Entry` and `VacantEntry` types for this case, while `OccupiedEntry` can remain shared.

Reply:
Here's a separate `FindEntry` with `AbsentEntry` (not firm on that name):
https://github.com/Amanieu/hashbrown/compare/hashtable...cuviper:hashbrown:hashtable-find_entry

Reply:
I think that the best thing for now is to make this `Result<OccupiedEntry, AbsentEntry>`, with the understanding that once NLL is fixed this will be changed to `Option<OccupiedEntry>`.

In the case of `find_entry`, we specifically want to use `find` instead of `find_or_find_insert_slot`, since the former is slightly faster.